### ISSUE TITLE
Working with Tls

### ISSUE THREAD

**Issue by yahalom** [gahr/retcl#13]
Working with Tls is less clear:
1. example in docs does not make it clear that the option are either or:
retcl create r ?-noconnect? ?host? ?port?
you get:
wrong # args: must be "?host port?" or "-noconnect"

2. To work with Tls is quite ugly:
retcl create r -noconnect
r +tls
r connect

it was much better if it could be:

retcl create r ?host? ?port? +tls

**Comment by gahr** [comment]
I don't disagree, the API for TLS is new and not very well thought through. I kind of added it in a hurry. At the time, I had evaluated the possibility to add a parameter to the constructor, but I thought it would be neater to keep it separate. The downside of course is that you have to create the object in `noconnect` mode. The upside is that we keep things separated. My concern is that - as the library grows - more and more parameters would end up in the constructor.

Now, having heard your feedback, I think I have one data-point telling me that perhaps my reasoning wasn't flawless.

Another idea is to add the possibility to specify all `-<flag>` / `+<flag>` options in the constructor.. at least this would be consistent.. you would be able to call, e.g., `retcl create r <host> <port> -keepcache +async +tls`. How does that sound?

Edit: on a second (third? fourth?) thought, it *would* get messy as soon as we wanted to also pass TLS parameters in the constructor..

`retcl create r <host> <port> -keepCache +tls -cafile /etc/redis/ca.crt -certfile /etc/redis/redis.crt -keyfile /etc/redis/redis.key -async`.

See how the TLS specific options and the retcl options would get all mixed up. I don't think I like it.

**Comment by gahr** [comment]
I *do* get the point that the docs could be clearer. Does [this version](https://code.ptrcrt.ch/retcl/doc/c670d128d95825c7/docs/index.html#_construction) improve clarity in your opinion?

**Comment by yahalom** [comment]
Yes, very clear.

Thanks

**Comment by yahalom** [comment]
Yes, having this options as part of the standard commands makes sense and
makes the code clear.

I do like the way the code is written with the 'unknown' method that lets
you be open to future redis changes. Well done.

It was very fast and easy to understand the package and start using it.

Thanks for sharing and supporting the package.

Yahalom

**Comment by yahalom** [comment]
I do not think this is messy:

retcl create r <host> <port> -keepCache +tls -cafile /etc/redis/ca.crt -certfile /etc/redis/redis.crt -keyfile /etc/redis/redis.key -async.

since the default behavior in object creation is to make connection to Redis, it makes sense to allow to send all needed connection parameters.
if this is not supported will it be:
retcl create r -noconnect
r +tls
r -cafile /etc/redis/ca.crt
r -keyfile /etc/redis/redis.key
r -async.
r connect

for me it look almost childish in its one config per row. I prefer the Tclcurl way:
$curlHandle configure -url $url -bodyvar html -post 1 -postfields $data  -sslversion tlsv1 -sslverifyhost 1

**Comment by gahr** [comment]
> I do not think this is messy:
> 
> ```
> retcl create r -keepCache +tls -cafile /etc/redis/ca.crt -certfile /etc/redis/redis.crt -keyfile /etc/redis/redis.key -async
> ```
...

> if this is not supported will it be:
> ```
> retcl create r -noconnect
> r +tls
> r -cafile /etc/redis/ca.crt
> r -certfile /etc/redis/redis.crt
> r -keyfile /etc/redis/redis.key
> r -async
> r connect
> ```

Now, I think this proves my point about it being messy: see that you are conflating two disjoint sets of options: `-keepCache`, `+tls`, and `-async` are Retcl options, while `-cafile`, `-certfile`, and `-keyfile` are TLS options, totally unknown to Retcl.

It just so happens that the `+tls` command to Retcl [accepts arbitrary parameters](https://github.com/gahr/retcl/blob/eb736260bce5bf53547c5b7de135e5ef7a8ec6a2/retcl.tm#L303-L304), and these are passed as-is to the [tls::socket command on connection](https://github.com/gahr/retcl/blob/eb736260bce5bf53547c5b7de135e5ef7a8ec6a2/retcl.tm#L216).

I don't want to make TLS options first-class citizens, (i.e., known) to Retcl. I want them to be kept as pass-through and transparent.

So, really, what we're looking for is something like this:

```
retcl create r host port -keepCache {+tls -cafile /etc/redis/ca.crt -certfile /etc/redis/redis.crt -keyfile /etc/redis/redis.key} -async
```

Where each additional argument after `host` and `port` is expanded to the corresponding Retcl call. This would be equivalent to:

```
retcl create r
r -keepCache
r +tls -cafile /etc/redis/ca.crt -certfile /etc/redis/redis.crt -keyfile /etc/redis/redis.key
r -async
r connect host port
```

I think I like the idea of this grouping of arguments, and this is what I'm likely going to implement. I just need to think over it some more :)

**Comment by gahr** [comment]
One complication is that `retcl create r -async` could be interpreted as either turning off `async` mode or using `-async` as a host name (a hyphen is a valid character in host names). Back to the drawing board.

**Comment by cgmacleod** [comment]
> One complication is that `retcl create r -async` could be interpreted as either turning off `async` mode or using `-async` as a host name (a hyphen is a valid character in host names). Back to the drawing board.

While it is legal, I don't think I've ever seen a hostname starting with "-".  It makes sense to guard against this though, so I would suggest the same approach as many Tcl commands already use: take "--" as marking the end of any options.

**Comment by gahr** [comment]
I just learned that hostnames can't start with a `-` (thank stu!). So I guess I can make it:

```
retcl create r ?host port? ?options?
retcl create r ?-noconnect? ?options?
```

Example:

```
retcl create r -keepCache {+tls -cafile /etc/redis/ca.crt -certfile /etc/redis/redis.crt -keyfile /etc/redis/redis.key} -async
```

**Comment by gahr** [comment]
I have implemented this, see the updated [docs](https://code.ptrcrt.ch/retcl/doc/4b1a087054b62a69/docs/index.html#_construction) and [tests](https://code.ptrcrt.ch/retcl/file?ci=4b1a087054b62a69&name=test/002-construct.test&ln=46-71).

**Comment by yahalom** [comment]
Looks great!

**Comment by gahr** [comment]
Fixed in 0.6.0.
